src/meta/src/stream/stream_graph.rs

```diff
@@ -321,8 +321,42 @@ impl StreamActorBuilder {
                 .flat_map(|(_, StreamActorUpstream { actors, .. })| actors.0.iter().copied())
                 .map(|x| x.as_global_id())
                 .collect(), // TODO: store each upstream separately
-            same_worker_node_as_upstream: self.chain_same_worker_node
-                || self.upstreams.values().any(|u| u.same_worker_node),
+            colocated_upstream_actor_id: if self.chain_same_worker_node {
+                if self.upstreams.is_empty() {
+                    None
+                } else {
+                    Some(ColocatedActorId {
```

Comment:
I've been feeling lately that this field of `Actor` seems a little strange. Actually, all `Chain`s are placed in the same parallel unit with the upstream actor, which is a result of some manual judgments in the code but is theoretically implied by the `NoShuffle` exchange.

I'm considering merging the scheduler logic into one-phase (instead of current `GraphBuilder` and then `Scheduler`, which is a result of not having consistent hash in old days), and directly generating the physical distribution that meets the requirements. I've tried solving this with some datalog-like representation. A trivial example can be found [here](https://github.com/BugenZhao/rust-playground/blob/master/src/bin/crepe.rs).

Reply:
The motivation or the requirement of this change is to handle the scheduling of the stream lookup node. Stream lookup node has two inputs, one is the outer side, the other is the inner side(the lookup side). We need to schedule the stream lookup node together with the inner side only instead of the outer side, so we need to specify which upstream actor currently. If we can imply this scheduling  logic directly from the `NoShuffle` exchange, it could be perfect.